### PR TITLE
Handle incomplete sequences at end of input

### LINKED ISSUE #67937: JSONCompact output format can emit unparsable JSON
**Describe what's wrong**

The `JSONCompact` format can produce unparsable JSON when the last value output ends with a byte between `0xf0` and `0xf7`.

**Does it reproduce on the most recent release?**

Yes

**How to reproduce**

 Running the query:

```
SELECT unhex('f0') FORMAT JSONCompact
```

Outputs
```
{
	"meta":
	[
		{
			"name": "unhex('f0')",
			"type": "String"
		}
	],

	"data":
	[
		["�
	],

	"rows": 1,

	"statistics":
	{
		"elapsed": 0.000510458,
		"rows_read": 1,
		"bytes_read": 1
	}
}
```

See the `data` section

[See in fiddle.clickhouse.com](https://fiddle.clickhouse.com/59006fc2-0b3b-4d91-a516-2c97766026a0).

**Expected behavior**

Parsable JSON is emitted

### PR DESCRIPTION
**Pull request by mwoenker**
### Changelog category (leave one):
- Improvement

### Changelog entry (a user-readable short description of the changes that goes to CHANGELOG.md):

* Improve unicode encoding in JSON output formats. Ensures that valid JSON is generated in the case of certain byte sequences in the result data.

Closes #67937

When producing JSON output, ClickHouse needs to serialize arbitrary strings that may have any sequence of bytes in them. To ensure that valid UTF-8 is produced, the JSON output formats inherit from the `WriteBufferValidUTF8` class, which filters the output, replacing UTF-8 errors with a replacement character `U+FFFD`.

Normally if the text contains a byte beginning a multi-byte sequence but the sequence is not valid, it skips the first byte, replacing it with a replacement character and continues encoding from there. If there is a byte beginning a multi byte sequence, but the text ends before that sequence should complete, the code simply writes a replacement character and stops the output there.

This causes this query:

```
SELECT unhex('f0') FORMAT JSONCompact;
```

to produce unparsable JSON with a data section looking like this:

```
	"data":
	[
		["�
	],
```

This PR changes the behavior to treat incomplete UTF-8 sequences the same as invalid UTF-8 sequences The first byte is replaced with the replacement character and encoding continues with the next byte.

### Documentation entry for user-facing changes

- [ ] Documentation is written (mandatory for new features)

> Information about CI checks: https://clickhouse.com/docs/en/development/continuous-integration/

#### CI Settings (Only check the boxes if you know what you are doing):
- [ ]  Allow: All Required Checks
- [ ]  Allow: Stateless tests
- [ ]  Allow: Stateful tests
- [ ]  Allow: Integration Tests
- [ ]  Allow: Performance tests
- [ ]  Allow: All Builds
- [ ]  Allow: batch 1, 2 for multi-batch jobs
- [ ]  Allow: batch 3, 4, 5, 6 for multi-batch jobs
---
- [ ]  Exclude: Style check
- [ ]  Exclude: Fast test
- [ ]  Exclude: All with ASAN
- [ ]  Exclude: All with TSAN, MSAN, UBSAN, Coverage
- [ ]  Exclude: All with aarch64, release, debug
---
- [ ]  Do not test
- [ ]  Woolen Wolfdog
- [ ]  Upload binaries for special builds
- [ ]  Disable merge-commit
- [ ]  Disable CI cache

####

### CODE REVIEW

**Comment by robot-clickhouse-ci-1** [comment]
*This is an automated comment for commit 704ec8dea6b4fa583689508e4362644517321005 with description of existing statuses. It's updated for the latest CI running*

[❌ Click here](https://s3.amazonaws.com/clickhouse-test-reports/67938/704ec8dea6b4fa583689508e4362644517321005/ci_running.html) to open a full report in a separate page

<table>
<thead><tr><th>Check name</th><th>Description</th><th>Status</th></tr></thead>
<tbody>
<tr><td>Style check</td><td>Runs a set of checks to keep the code style clean. If some of tests failed, see the related log from the report</td><td><a href="https://s3.amazonaws.com/clickhouse-test-reports/67938/704ec8dea6b4fa583689508e4362644517321005/style_check.html">❌ failure</a></td></tr>
<tbody>
</table>

**Comment by nikitamikhaylov** [comment]
@mwoenker Let's add a test? To do it, please: 
- Go to `tests/queries/0_stateless`
- Run `./add-test incomplete-sequence-bla-bla-bla`
- Write some SQL code (in case if you need something else, there is also an option to write `.sh` test)
- Run `clickhouse-client -nm < <your_test_name>.sql > <your_test_name>.reference`
- Verify using `tests/clickhouse-test <your_test_name>`

**Comment by mwoenker** [comment]
Test created

**Comment by pamarcos** [comment]
Two nitpick comments :pray: 

1. Please use only underscores for the name of the test for consistency with the rest: `03221_incomplete-utf8-sequence.reference` -> `03221_incomplete_utf8_sequence.reference`
2. I'm not sure the test is testing your code modifications. I've tried running the test with 24.5.3.5, 24.6.2.17 and 24.7.2.13 and it was successful in all of them. I've also set breakpoints in the function you modified and when running the same query from the test (`SELECT unhex('f0') FORMAT JSONCompact`), no breakpoint is hit. Maybe I'm missing something :thinking:

**Review by pamarcos — APPROVED** [review]
Code LGTM :rocket: 

Two minor things:
1. The test name should use underscores instead of hyphen for consistency
2. Comments should use three slashes `///` instead of two `//` according to our coding style guideline

File: src/IO/WriteBufferValidUTF8.cpp [hunk 1]
```diff
@@ -149,9 +149,34 @@ void WriteBufferValidUTF8::finalizeImpl()
     /// Write all complete sequences from buffer.
     nextImpl();
 
-    /// If unfinished sequence at end, then write replacement.
+    /// Handle remaining bytes if we have an incomplete sequence
     if (working_buffer.begin() != memory.data())
-        putReplacement();
+    {
+        char * p = memory.data();
```
Comment (alexey-milovidov) [inline src/IO/WriteBufferValidUTF8.cpp:155]: `const char *`?

**Comment by alexey-milovidov** [comment]
Let's also change the category to `Improvement`, because it is a great improvement!

**Comment by alexey-milovidov** [comment]
(then take the PR template from `.github` directory and put the description into the changelog entry)

**Comment by mwoenker** [comment]
Changed to Improvement, and changed github label accordingly.

Let me know if this short description sounds good.

**Comment by clickgapai** [comment]
Hi @mwoenker @pamarcos — while reviewing this PR I found the following:

- **Bug (P1):** `finalizeImpl` loop uses stale `pos` after `nextImpl` shifts data, causing duplicate output — https://github.com/ClickHouse/ClickHouse/issues/104982

Happy to discuss — close anything that's wrong or already addressed.

**Pull request merged**